### PR TITLE
file scope API features

### PR DESCRIPTION
**Pull request by williballenthin**
adds support for extracting `API` features at file scope to indicate statically linked routines, such as extracted via FLIRT. see idea/discussion in #567

### CODE REVIEW

**Comment by williballenthin** [comment]
enables rules such as:

```yml
rule:
  meta:
    name: encrypt data using AES via CryptoPP
    namespace: data-manipulation/encryption/aes
    author: william.ballenthin@fireeye.com
    scope: file
    att&ck:
      - Defense Evasion::Obfuscated Files or Information [T1027]
    mbc:
      - Defense Evasion::Obfuscated Files or Information::Encryption-Standard Algorithm [E1027.m05]
      - Cryptography::Encrypt Data::AES [C0027.001]
    examples:
      - 8BA66E4B618FFDC8255F1DF01F875DDE6FD0561305D9F8307BE7BB11D02AE363
  features:
    - or:
      - api: "?ProcessAndXorBlock@Enc@Rijndael@CryptoPP@@UBEXPBE0PAE@Z"
```

File: capa/features/extractors/smda/file.py [hunk 1]
```diff
@@ -79,8 +79,8 @@ def extract_file_import_names(smda_report, file_path):
         library_name = imported_library.name.lower()
         library_name = library_name[:-4] if library_name.endswith(".dll") else library_name
         for func in imported_library.entries:
+            va = func.iat_address + smda_report.base_addr
             if func.name:
-                va = func.iat_address + smda_report.base_addr
```
Comment (williballenthin) [inline capa/features/extractors/smda/file.py:83]: unrelated fix

File: capa/features/extractors/smda/file.py [hunk 2]
```diff
@@ -125,7 +135,6 @@ def extract_features(smda_report, file_path):
     """
 
     for file_handler in FILE_HANDLERS:
-        result = file_handler(smda_report, file_path)
```
Comment (williballenthin) [inline capa/features/extractors/smda/file.py:128]: unrelated fix

File: capa/features/extractors/ida/file.py [hunk 3]
```diff
@@ -143,6 +143,16 @@ def extract_file_strings():
             yield String(s.s), (seg.start_ea + s.offset)
 
 
+def extract_file_library_functions():
+    """
+    extract the names of statically-linked library functions.
+    """
+    if False:
+        # using a `yield` here to force this to be a generator, not function.
+        yield NotImplementedError("SMDA doesn't have library matching")
```
Comment (mr-tz) [inline capa/features/extractors/ida/file.py:152]: this is `IDA`, can't we its function ID here?

**Review by mr-tz — COMMENTED** [review]
looks good, just one question on doing this in IDA

**Comment by mr-tz** [comment]
we also need to update the documentation

**Comment by williballenthin** [comment]
i have pending documentation changes but they're in `capa-rules` rather than this repo.

**Comment by williballenthin** [comment]
im not convinced that `API` is the right term here. alternatively options include:

  - `name`
  - `symbol`
  - `function-name`

**Comment by mr-tz** [comment]
Would love to hear what others think on how we should name this feature.

@re-fox @johnk3r @recvfrom @r0ny123

**Comment by r0ny123** [comment]
I think `function-name` (as proposed by @williballenthin ) or `library functions` would better reflect the idea.

File: capa/features/extractors/ida/file.py [hunk 4]
```diff
@@ -143,6 +143,16 @@ def extract_file_strings():
             yield String(s.s), (seg.start_ea + s.offset)
 
 
+def extract_file_library_functions():
+    """
+    extract the names of statically-linked library functions.
+    """
+    if False:
```
Comment (mike-hunhoff) [inline capa/features/extractors/ida/file.py:150]:
```suggestion
    for ea in idautils.Functions():
	    if idaapi.get_func(ea).flags & idaapi.FUNC_LIB:
		    name = idaapi.get_name(ea)
		    yield capa.features.insn.API(name), ea
```

Comment (mike-hunhoff) [reply capa/features/extractors/ida/file.py:150]: I think the formatting is off with my suggestion but we could use this code to identify library function names as see by IDA.

**Review by mike-hunhoff — COMMENTED** [review]
Looking good! LGTM once we decide on a name for the new feature.

**Comment by recvfrom** [comment]
I think I like `function-name` the best from the options above, although it's a bit long (would just using `function` be too ambiguous?)

**Comment by williballenthin** [comment]
lets go with `function-name` thanks to the feedback above!

**Comment by williballenthin** [comment]
ready for re-review. feature is named `function-name` and applies at file scope.

File: capa/features/__init__.py [hunk 5]
```diff
@@ -7,11 +7,11 @@
 # See the License for the specific language governing permissions and limitations under the License.
```
Comment (williballenthin) [inline capa/features/__init__.py:7]: unrelated fixes in this file

File: capa/features/extractors/viv/file.py [hunk 6]
```diff
@@ -80,6 +83,16 @@ def extract_file_strings(vw, file_path):
         yield String(s.s), s.offset
 
 
+def extract_file_function_names(vw, file_path):
+    """
+    extract the names of statically-linked library functions.
+    """
+    for va in sorted(vw.getFunctions()):
```
Comment (williballenthin) [inline capa/features/extractors/viv/file.py:90]: notably, wrt #583, this requires an analysis pass by viv to identify functions

**Review by mr-tz — APPROVED** [review]
awesome!

**Comment by williballenthin** [comment]
hm need to update tests to xfail SMDA due to not having FLIRT

File: tests/fixtures.py [hunk 7]
```diff
@@ -314,179 +314,186 @@ def parametrize(params, values, **kwargs):
...
+        # function/characteristic(calls to)
+        ("mimikatz", "function=0x40105D", capa.features.Characteristic("calls to"), True),
+        # before this we used ambiguous (0x4556E5, False), which has a data reference / indirect recursive call, see #386
+        ("mimikatz", "function=0x456BB9", capa.features.Characteristic("calls to"), False),
+        # file/function-name
+        ("pma16-01", "file", capa.features.file.FunctionName("__aulldiv"), True),
+    ],
+    # order tests by (file, item)
+    # so that our LRU cache is most effective.
+    key=lambda t: (t[0], t[1]),
```
Comment (williballenthin) [inline tests/fixtures.py:495]: @mr-tz FYI

**Review by Ana06 — COMMENTED** [review]
This looks great! 💘 Could you squash the merge/pep8 commits? (or can we merge this with the `squash and merge` option?)
Please

File: tests/fixtures.py (hunk 7 above)
Comment (mr-tz) [reply tests/fixtures.py:495]: nice!

**Comment by williballenthin** [comment]
#585 SMDA not working on Windows for two tests.
